### PR TITLE
Add a TriggerUponDiscoveringTile unique type

### PR DESCRIPTION
**Pull request by karmaBonfire**
These changes add a `TriggerUponDiscoveringTile` unique of target `UnitTriggerCondition`. It allows to define actions that occur when a unit discovers a previously fogged tile and triggers for each such tile.

It is a modding-related feature since AFAIK there are no relevant mechanics in Civ5.

I have tested following uniques:
 - `This Unit gains [5] XP <upon discovering a [Water] tile>`
 - `This Unit gains [5] XP <with [50]% chance> <upon discovering a [Water] tile>`

This is my first Unciv PR and my first Kotlin PR - reviews are welcome and encouraged!

### CODE REVIEW

File: core/src/com/unciv/logic/map/mapunit/MapUnit.kt [hunk 1]
```diff
@@ -343,8 +344,23 @@ class MapUnit : IsPartOfGameInfoSerialization {
...
+            }
+            for (tile in newlyExploredTiles) {
+                // Include tile in the state for correct RNG seeding
+                val state = StateForConditionals(civInfo=civ, unit=this, tile=tile);
+                for (unique in getTriggeredUniques(UniqueType.TriggerUponDiscoveringTile, state)) {
+                    if (unique.conditionals.any { it.type == UniqueType.TriggerUponDiscoveringTile
+                        && tile.matchesFilter(it.params[0], civ) })
+                        UniqueTriggerActivation.triggerUnitwideUnique(unique, this)
+                }
+            }
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/mapunit/MapUnit.kt:360]:
```suggestion
            if (newlyExploredTiles.isNotEmpty){
                val unfilteredTriggeredUniques = getTriggeredUniques(UniqueType.TriggerUponDiscoveringTile, StateForConditionals.IgnoreConditionals)
                for (tile in newlyExploredTiles) {
                    // Include tile in the state for correct RNG seeding
                    val state = StateForConditionals(civInfo=civ, unit=this, tile=tile);
                    for (unique in unfilteredConditionals) {
                        if (unique.conditionals.any { it.type == UniqueType.TriggerUponDiscoveringTile
                            && tile.matchesFilter(it.params[0], civ) }
                                && unique.conditionalsApply(state))
                            UniqueTriggerActivation.triggerUnitwideUnique(unique, this)
                    }
                }
            }
```

I think this is what SomeTrog was aiming for, tbh my first thought was also to get triggerables once

**Review by yairm210 — APPROVED** [review]
All in all VERY impressive first PR and a fantastic flavor for unique explorer types!

File: core/src/com/unciv/logic/map/mapunit/MapUnit.kt (hunk 1 above)
Comment (yairm210) [reply core/src/com/unciv/logic/map/mapunit/MapUnit.kt:360]: You're right, it needs a .toList() :)

**Comment by karmaBonfire** [comment]
@SomeTroglodyte @yairm210 Thanks, I really appreciate your feedback! Glad you like this feature. I just wanted to make a mod but found out there's no triggers I need, so I decided to dive deeper and implement it :)

File: core/src/com/unciv/logic/map/mapunit/MapUnit.kt (hunk 1 above)
Comment (karmaBonfire) [reply core/src/com/unciv/logic/map/mapunit/MapUnit.kt]: That's fair, given that this trigger is pretty niche, the game would calculate newlyExploredTiles far more often than needed. I've implemented a fix with both your suggestions, thanks again for your comments!

**Comment by karmaBonfire** [comment]
> We have no way of telling modders this conditional will only work on units, right?

Well, this trigger is listed in UnitTriggerCondition category, so I suppose it's clear enough it only works for units

**Comment by yairm210** [comment]
Unciv is currently a good *emergent storytelling* tool but a very, VERY bad *structured narrative* tool... I think anyone who wants something like that is best served going to another game

**Pull request merged**